src/raster/buffer.rs

```diff
@@ -0,0 +1,305 @@
+/// assert_eq!(max.round(), 168.0);
+///
+/// // Mutate the buffer and observe the difference:
+/// let mut buf = buf;
+/// buf[center] = 1e10;
+/// let (min, max) = min_max(&buf);
+/// assert_eq!(max, 1e10);
+///
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub struct Buffer<T> {
+    shape: (usize, usize),
+    data: Vec<T>,
+}
+
+impl<T: GdalType> Buffer<T> {
+    /// Construct a new buffer from `size` (`(cols, rows)`) and `Vec<T>`.
+    ///
+    /// # Notes
+    /// The elements of `shape` are in reverse order from what is used in `ndarray`.
+    ///
+    /// # Panics
+    /// Will panic if `size.0 * size.1 != data.len()`.
+    pub fn new(shape: (usize, usize), data: Vec<T>) -> Self {
+        assert_eq!(
+            shape.0 * shape.1,
+            data.len(),
+            "shape {}*{}={} does not match length {}",
+            shape.0,
+            shape.1,
+            shape.0 * shape.1,
+            data.len()
+        );
+        Buffer { shape, data }
+    }
+
+    /// Destructures `self` into constituent parts.
+    pub fn into_shape_and_vec(self) -> ((usize, usize), Vec<T>) {
+        (self.shape, self.data)
+    }
+
+    /// Gets the 2-d shape of the buffer.
+    //
+    /// Returns `(cols, rows)`
+    ///
+    /// # Notes
+    /// The elements of `shape` are in reverse order from what is used in `ndarray`.
+    pub fn shape(&self) -> (usize, usize) {
+        self.shape
+    }
+
+    /// Get a slice over the buffer contents.
+    pub fn data(&self) -> &[T] {
+        self.data.as_slice()
+    }
+
+    /// Get a mutable slice over the buffer contents.
+    pub fn data_mut(&mut self) -> &mut [T] {
+        self.data.as_mut_slice()
+    }
+
+    /// Get the number of elements in the buffer
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Determine if the buffer has no elements.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    #[cfg(feature = "ndarray")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "ndarray")))]
+    /// Convert `self` into an [`ndarray::Array2<T>`].
+    pub fn to_array(self) -> crate::errors::Result<Array2<T>> {
+        // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
+        Ok(Array2::from_shape_vec(
+            (self.shape.1, self.shape.0),
+            self.data,
+        )?)
+    }
+
+    #[inline]
+    fn vec_index_for(&self, coord: (usize, usize)) -> usize {
+        if coord.0 >= self.shape.0 {
+            panic!(
+                "index out of bounds: buffer has {} columns but row {} was requested",
+                self.shape.0, coord.0
+            );
+        }
+        if coord.1 >= self.shape.1 {
+            panic!(
+                "index out of bounds: buffer has {} rows but row {} was requested",
+                self.shape.1, coord.1
+            );
+        }
+        coord.0 * self.shape.0 + coord.1
```

Comment:
I think we should move the `panic!`s into another function with `#[cold] #[inline(never)]`. It's all right if we check the coordinates twice (both here and in that function).

Reply:
Actually, I think a single `panic!` that prints the index and shape should be enough. The user can figure it out.